### PR TITLE
speed sam.ParseCigar

### PR DESCRIPTION
**Pull request by brentp**
```
benchmark                 old ns/op     new ns/op     delta
BenchmarkParseCigar-4     18076         3956          -78.11%
```
I'm using sam.ParseCigar to parse the cigar in SA tags in bams from oxford nanopore and the cigar strings are very long. This improves the performance of the full program somewhat. 
It's specialized, but it is safe since there are already checks the '0' <= c <= '9' for all characters we're converting to int.

### CODE REVIEW

**Comment by brentp** [comment]
done.

File: sam/cigar.go [hunk 1]
```diff
@@ -197,6 +196,21 @@ func init() {
 	}
 }
 
+var powers = []int{1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000}
```
Comment (kortschak) [inline sam/cigar.go]: Please use e notation.

File: sam/cigar.go [hunk 2]
```diff
@@ -197,6 +196,21 @@ func init() {
...
+
+// atoi is an internal function; bytes in b are known to be digits.
```
Comment (kortschak) [inline sam/cigar.go]: `// atoi returns the integer interpretation of b which must be an ASCII decimal number representation.`

File: sam/cigar.go [hunk 3]
```diff
@@ -197,6 +196,21 @@ func init() {
...
+func atoi(b []byte) (int, error) {
+	n := 0
+	k := len(b) - 1
+	for i, v := range b {
+		n += int(v-'0') * powers[k-i]
+	}
+	if n >= 1<<28 || n < 0 {
```
Comment (kortschak) [inline sam/cigar.go]: `n < 0 || 1<<28 <= n`. This says n outside 0 to 1<<28 very clearly, so is preferred.

There is no test for invalid CIGAR handling. I'll send a PR for that.

File: sam/cigar.go [hunk 4]
```diff
@@ -197,6 +196,21 @@ func init() {
...
+		return n, fmt.Errorf("sam: cigar-op size out of bounds [0, 1<<28): %d", n)
```
Comment (kortschak) [inline sam/cigar.go]: `fmt.Errorf("sam: invalid cigar operation count: %q", b)`. Ideally, I'd like to have `i` here as well so the start can be pointed to in the error, so `fmt.Errorf("sam: invalid cigar operation count: %q at %d", b, i)`.

File: sam/cigar.go [hunk 5]
```diff
@@ -211,11 +225,11 @@ func ParseCigar(b []byte) (Cigar, error) {
 	for i := 0; i < len(b); i++ {
 		for j := i; j < len(b); j++ {
 			if b[j] < '0' || '9' < b[j] {
-				n, err = strconv.Atoi(string(b[i:j]))
+				op = cigarOpTypeLookup[b[j]]
```
Comment (kortschak) [inline sam/cigar.go]: Please put this back where it was.

**Comment by brentp** [comment]
I made those changes.

File: sam/cigar.go [hunk 6]
```diff
@@ -197,6 +196,21 @@ func init() {
 	}
 }
 
+var powers = []int{1, 1e1, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8}
```
Comment (kortschak) [inline sam/cigar.go]: s/1/1e0/ (regularity trumps low character count).

File: sam/sam_test.go [hunk 7]
```diff
@@ -751,3 +751,14 @@ func (s *S) TestIssue32(c *check.C) {
...
+
+func BenchmarkParseCigar(b *testing.B) {
+	cig := []byte("69S17M5I30M1D45M1D23M5I14M2I4M1I10M2D7M1D6M14I33M1D6M1I7M1I18M1I8M1D4M1D4M2D57M1D21M1D6M1I14M1I7M1I3M1I9M1D3M1D7M1D37M1D9M1I5M1I15M4I12M1D10M1I10M1D8M1D26M7I12M1D20M1I36M1I22M3D8M1I23M1I13M2D10M1D12M1I15M6D4M1D4M1D1M2D5M1D3M17D1M1D13M3D7M1I29M2I9M1D2M4D7M2D8M5D3M1D1M1D23M1D10M6D19M3I24M1D8M1I11M6D14M1I5M8I12M1D8M2D5M2D2M1D23M1D11M4I35M2I19M1I4M1D13M7I33M1D21M3D2M1D9M4I19M1I14M1D7M1I41M1D23M3I18M1I6M1I13M1D9M1D1M1D20M1D23M5D8M1I13M2I11M1D78M2I18M10D9M2D10M1D10M2I6M1D3M1D21M2I7M1D7M2I12M1D20M2D18M1I12M1D8M4D18M1D6M1D20M1D14M1I1M2I23M1I10M1D7M1I15M1D4M1I9M1D11M1D12M1I8M1D21M1I13M2I59M1D12M1D18M1D13M1D22M1D13M1I19M1D13M1D19M1I11M2I27M2D10M1D17M6D13M2D17M1D13M1D19M1I3M1D13M2I33M1I26M2D9M2I21M2D10M1D36M1D32M5I23M1D13M2D17M1I14M2I24M1I5M2I8M2I24M2I9M1D7M1D2M1D15M3I19M1I2M1D3M1I7M1D5M2D24M5I1M4I33M1I13M3I34M1I2M1I23M1D3M2I8M1I5M5S")
+	for i := 0; i < b.N; i++ {
+		_, err := ParseCigar(cig)
+		if err != nil {
+			panic(err)
+		}
+	}
+
```
Comment (kortschak) [inline sam/sam_test.go]: Delete blank line.

**Comment by brentp** [comment]
done

**Comment by kortschak** [comment]
Thanks you.

**Pull request merged**